### PR TITLE
Database modification

### PR DESCRIPTION
**Pull request by MateuszSzalas**
Adds dialog for input of parts missing from database and then insert them into database

### CODE REVIEW

File: app/excel-parser/database.ts [hunk 1]
```diff
@@ -15,10 +17,17 @@ export async function getPartPropertiesFromDatabase(partNr: string): Promise<{sa
              // eslint-disable-next-line @typescript-eslint/naming-convention
              (_, row: {report_nr: string; sap_nr: string; sap_name: string}) => {
             if (row === undefined) {
-                resolve({sapNr: 'Not in db', sapName: 'Not in db'});
```
Comment (Jack53416) [inline app/excel-parser/database.ts:18]: That's not a good practice regarding an Error handling. You are introducing some magic value in one field. That's confusing and prone to errors. 

Use reject callback and typical error handling with exceptions. 

TLDR guide to exceptions:
1. Create your own base exception class to handle most generic errors coming from your library
2. Subclass this exception for more specific ones like for instance PartNotFoundError
3. Encapsulate necessary properties within exception class, in this case 'sapNr'
4. Handle exceptions in try catch clause in upper context


Note that in JS catch clause won't support types. So you manually have to check whether the exception is an instance of desired one and re-throw one that is not handled

You can read more:
https://fettblog.eu/typescript-typing-catch-clauses/
https://javascript.info/custom-errors

File: app/excel-parser/save-report.ts [hunk 2]
```diff
@@ -27,27 +29,38 @@ async function createPartWorkbook(fileName: string, yearShort: string): Promise<
...
   }
 
-  await partWorkbook.xlsx.writeFile(`${statisticsPath}/${fileName}.xlsx`);
+  if (partData.machine !== fileName.slice(0, 6)) {
+    monthSheet.getCell(`${partStatisticColumns.totalPartsProduced}${row}`).note = partData.machine;
+  }
+
+  // set edited worksheet to be active when open file manually
+  partWorkbook.views = [
+    {
```
Comment (Jack53416) [inline app/excel-parser/save-report.ts]: Format this object properly, one property - one line

File: src/app/core/services/electron/electron.service.ts [hunk 3]
```diff
@@ -66,6 +67,16 @@ export class ElectronService {
     return result;
   }
 
+  async addPartToDatabase(updatedParts: UpdatedPart[]) {
+    const partsToUpdate: UpdatedPart[] = [];
+
+    for (const part of updatedParts) {
+      if (part.addToDatabase) {partsToUpdate.push(part);}
```
Comment (Jack53416) [inline src/app/core/services/electron/electron.service.ts]: Don't  format ifs like that, always use multiple lines for its body

File: src/app/core/services/electron/electron.service.ts [hunk 4]
```diff
@@ -66,6 +67,16 @@ export class ElectronService {
     return result;
   }
 
+  async addPartToDatabase(updatedParts: UpdatedPart[]) {
+    const partsToUpdate: UpdatedPart[] = [];
+
+    for (const part of updatedParts) {
```
Comment (Jack53416) [inline src/app/core/services/electron/electron.service.ts]: Use filter operator on Array. In general you are using way too much for loops than its necessary. Embrace built in functions. But also bear in mind consequences as well. Each array pass increases a big O complexity. If you can do more than one thing in a single pass use for loop otherwise there is no point.

File: src/app/parts-editor/components/dialog-database/dialog-database.component.html [hunk 5]
```diff
@@ -0,0 +1,20 @@
+<h1 i18n>Parts missing from database</h1>
```
Comment (Jack53416) [inline src/app/parts-editor/components/dialog-database/dialog-database.component.html:1]: Naming conventions of dialog can be better. This name is ambiguous. It doesn't tell dialogs purpose.  Its purpose is to add missing parts, so it would be much better to call it  MissingPartsDialog, PartListDialog, DbPartListDialog

File: src/app/parts-editor/components/dialog-database/dialog-database.component.html [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    <div *ngFor="let part of updatedParts().controls; let partIndex = index">
+      <div [formGroupName]="partIndex">
+        <input type="text" readonly formControlName="rowIndex" style="width: auto" />
+        <span i18n>Report number :</span>
+        <input type="text" readonly formControlName="nameReport" />
+        <span i18n>Sap name</span>
+        <input type="text" formControlName="nameSap" />
+        <span i18n>Sap number</span>
+        <input type="text" formControlName="numberSap" />
+        <span i18n>Add to databse</span>
```
Comment (Jack53416) [inline src/app/parts-editor/components/dialog-database/dialog-database.component.html]: Typo in word database

File: src/app/parts-editor/components/dialog-database/dialog-database.component.html [hunk 7]
```diff
@@ -0,0 +1,20 @@
+<h1 i18n>Parts missing from database</h1>
+
+<form [formGroup]="missingPartsForm" (ngSubmit)="onSubmit()">
+  <div formArrayName="updatedParts">
+    <div *ngFor="let part of updatedParts().controls; let partIndex = index">
+      <div [formGroupName]="partIndex">
+        <input type="text" readonly formControlName="rowIndex" style="width: auto" />
+        <span i18n>Report number :</span>
```
Comment (Jack53416) [inline src/app/parts-editor/components/dialog-database/dialog-database.component.html]: There is dedicated label html tag for labels. Do not use ordinary spans for that. Also it would be much better to use a table layout as labels are the same for each row

File: src/app/parts-editor/components/dialog-database/dialog-database.component.html [hunk 8]
```diff
@@ -0,0 +1,20 @@
+<h1 i18n>Parts missing from database</h1>
+
+<form [formGroup]="missingPartsForm" (ngSubmit)="onSubmit()">
```
Comment (Jack53416) [inline src/app/parts-editor/components/dialog-database/dialog-database.component.html]: NgSubmit is a part of FormsModule, not ReactivePartsModule. Also semantically submitting is not correct here. submitting a form means sending POST request to server. 

Do not use type submit on a button without clear reason. Use standard click handler on a button, and only close dialog when the underlying form control is valid

File: src/app/parts-editor/components/dialog-database/dialog-database.component.html [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  <div formArrayName="updatedParts">
+    <div *ngFor="let part of updatedParts().controls; let partIndex = index">
+      <div [formGroupName]="partIndex">
+        <input type="text" readonly formControlName="rowIndex" style="width: auto" />
```
Comment (Jack53416) [inline src/app/parts-editor/components/dialog-database/dialog-database.component.html]: Never use inline styles. Also width auto is its default value so this does nothing.

File: src/app/parts-editor/components/dialog-database/dialog-database.component.html [hunk 10]
```diff
@@ -0,0 +1,20 @@
+<h1 i18n>Parts missing from database</h1>
+
+<form [formGroup]="missingPartsForm" (ngSubmit)="onSubmit()">
+  <div formArrayName="updatedParts">
+    <div *ngFor="let part of updatedParts().controls; let partIndex = index">
+      <div [formGroupName]="partIndex">
```
Comment (Jack53416) [inline src/app/parts-editor/components/dialog-database/dialog-database.component.html]: This div is pointless , use the div above form that

File: src/app/parts-editor/components/dialog-database/dialog-database.component.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  missingPartsForm: FormGroup;
+
+  constructor(
+    public dialogRef: MatDialogRef<DialogDatabaseComponent>,
+    private formBuilder: FormBuilder,
+    @Inject(MAT_DIALOG_DATA) public missingParts: PartWorkbook
+  ) {}
+
+  ngOnInit() {
+    this.missingPartsForm = this.formBuilder.group({
```
Comment (Jack53416) [inline src/app/parts-editor/components/dialog-database/dialog-database.component.ts]: The top group is pointless. There is only one property there, just use array as a top level

File: src/app/parts-editor/components/dialog-database/dialog-database.component.ts [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+      this.addPart(part);
+    }
+  }
+
+  ngOnDestroy(): void {
+    this.destroy$.next();
+    this.destroy$.complete();
+  }
+
+  updatedParts(): FormArray {
```
Comment (Jack53416) [inline src/app/parts-editor/components/dialog-database/dialog-database.component.ts]: Use getter for that

File: src/app/parts-editor/components/dialog-database/dialog-database.component.ts [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    return this.missingPartsForm.get('updatedParts') as FormArray;
+  }
+
+  addPart(part: {[key: string]: Cell}) {
+    const newPart = this.formBuilder.group({
```
Comment (Jack53416) [inline src/app/parts-editor/components/dialog-database/dialog-database.component.ts]: Return this instead, and add it to array in upper context

File: src/app/parts-editor/components/dialog-database/dialog-database.component.ts [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    private formBuilder: FormBuilder,
+    @Inject(MAT_DIALOG_DATA) public missingParts: PartWorkbook
+  ) {}
+
+  ngOnInit() {
+    this.missingPartsForm = this.formBuilder.group({
+      updatedParts: this.formBuilder.array([]),
+    });
+
+    for (const part of this.missingParts) {
```
Comment (Jack53416) [inline src/app/parts-editor/components/dialog-database/dialog-database.component.ts]: Again use map array function for that

File: src/app/parts-editor/components/dialog-database/dialog-database.component.ts [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+      nameReport: part.articleNo.value,
+      nameSap: '',
+      numberSap:'',
+      addToDatabase: false
+    });
+
+    this.updatedParts().push(newPart);
+  }
+
+  newPart(partNumber: string): FormGroup {
```
Comment (Jack53416) [inline src/app/parts-editor/components/dialog-database/dialog-database.component.ts]: Function name should describe the action its doing. Do not use nouns for function names unless its a getter / setter and it mimics the property.  Here I suggest using createPartForm()

File: src/app/parts-editor/services/parts-data.service.ts [hunk 16]
```diff
@@ -5,12 +5,22 @@ import { BehaviorSubject, Observable, Subject } from 'rxjs';
...
 import { Cell } from '../models/cell';
+import { DialogDatabaseComponent } from '../components/dialog-database/dialog-database.component';
+import { MatDialog } from '@angular/material/dialog';
 
 export interface EditorsUiState {
   activeIndex: number | null;
   openedEditors: PartEditor[];
 }
 
+export interface UpdatedPart {
```
Comment (Jack53416) [inline src/app/parts-editor/services/parts-data.service.ts]: If its not strongly connected with this service it should be moved to separate file in models directory. Also probably it would be better to rename as PartToUpdate. Another approach for that crud models is to use crud operation as suffix ex PartUpdate, PartCreate etc

File: src/app/parts-editor/services/parts-data.service.ts [hunk 17]
```diff
@@ -124,6 +136,35 @@ export class PartsDataService {
...
+  private updateDatabase() {
+    const partsMissing = [];
+
+    for (const part of this.activeEditor.workbook) {
+      if (part.name.value === 'Not in db') {
+        partsMissing.push(part);
+      }
+    }
+
+    if (partsMissing.length > 0) {this.openDialogDatabase(partsMissing);}
```
Comment (Jack53416) [inline src/app/parts-editor/services/parts-data.service.ts]: No one line ifs

File: src/app/parts-editor/services/parts-data.service.ts [hunk 18]
```diff
@@ -124,6 +136,35 @@ export class PartsDataService {
     return this.getMockedData();
   }
 
+  private updateDatabase() {
+    const partsMissing = [];
+
+    for (const part of this.activeEditor.workbook) {
```
Comment (Jack53416) [inline src/app/parts-editor/services/parts-data.service.ts]: Again use filter function for that

**Review by Jack53416 — APPROVED** [review]
